aten/src/ATen/native/cuda/jit_utils.cpp

```diff
@@ -747,6 +771,17 @@ std::string generate_code(
     env.s("complex_body_string", "");
     env.s("complex_math_string", "");
   }
+  if (f_inputs_type == "std::complex<at::Half>" ||
+      result_type == "std::complex<at::Half>" || dynamic_casting) {
```

Comment:
Why is this not an `else if` in the `if` clause above? I find particularly weird the "or `dynamic_casting` part of this, as I do not fully understand how is that related to Half, and how it relates to the `else if (dynamic_casting)`  clause above. Even more, I don't know whether this is correct in the case that `f_input_type` and `result_type` are not Half, but `dynamic_casting == true`.

I think this should be split into two parts. The first one would modify
https://github.com/pytorch/pytorch/blob/4ebf4d49e9e96cc4e8f032db1a0afd4929ee1f37/aten/src/ATen/native/cuda/jit_utils.cpp#L748
to include the `std::complex<at::Half>` case.

The second part would go into the if-else above this one, and would add a `std::complex<at:.Half>` case to
https://github.com/pytorch/pytorch/blob/4ebf4d49e9e96cc4e8f032db1a0afd4929ee1f37/aten/src/ATen/native/cuda/jit_utils.cpp#L760-L761
and then do an if-else inside to pass `get_complex_body_string()` or add the `half_body` to it as well.

Note that these changes are not equivalent to the code currently implemented.

Reply:
Thanks for raising this point that this was a bit complex. Have updated/simplified the logic.

As for the `dynamic_casting == true` case, we require the definitions for all type for the Load and Store structures which cast the data before loading and storing.

https://github.com/pytorch/pytorch/blob/b09769992f83f94150eaef2ab9d03c37b36da159/aten/src/ATen/native/cuda/jit_utils.cpp#L266

https://github.com/pytorch/pytorch/blob/b09769992f83f94150eaef2ab9d03c37b36da159/aten/src/ATen/native/cuda/jit_utils.cpp#L260-L271

https://github.com/pytorch/pytorch/blob/b09769992f83f94150eaef2ab9d03c37b36da159/aten/src/ATen/native/cuda/jit_utils.cpp#L273-L281

Reply:
Nice refactor, but we can take it further. You can factor out the `traits` and the `complex` parts from `Half` into the condition
https://github.com/pytorch/pytorch/blob/610e18cc632ef51e20560dcb1af5b08a78587abd/aten/src/ATen/native/cuda/jit_utils.cpp#L761-L762

This way, each if-else statement sets just one subset of the `env` parameters, so it's easier to track what's set where.

Reply:
@Lezcano just to confirm, are you suggesting something like

```patch
--- a/aten/src/ATen/native/cuda/jit_utils.cpp
+++ b/aten/src/ATen/native/cuda/jit_utils.cpp
@@ -746,7 +746,10 @@ std::string generate_code(
     functor_args << "arg0[j], scalar_val";
   }
   env.s("args", functor_args.str());
-  if (f_inputs_type == "at::Half" || result_type == "at::Half" || dynamic_casting) {
+  if (f_inputs_type == "at::Half" || result_type == "at::Half" ||
+      f_inputs_type == "std::complex<at::Half>" ||
+      result_type == "std::complex<at::Half>" || dynamic_casting) {
+    // complex<Half> depends on complex<T> and Half dtypes.
     env.s("half_string", jiterator_half_support_literal);
   } else {
     env.s("half_string", "");
@@ -759,7 +762,9 @@ std::string generate_code(
   // the definition of complex math functions is only needed when the compute type is complex
   // but the definition of std::complex is needed for dynamic casting even if the compute type is not complex
   if (f_inputs_type == "std::complex<float>" || result_type == "std::complex<float>" ||
-      f_inputs_type == "std::complex<double>" || result_type == "std::complex<double>") {
+      f_inputs_type == "std::complex<double>" || result_type == "std::complex<double>" ||
+      f_inputs_type == "std::complex<at::Half>" || result_type == "std::complex<at::Half>") {
+    // complex<Half> depends on complex<T> and Half dtypes.
     env.s("traits_string", get_traits_string());
     env.s("complex_body_string", get_complex_body_string());
     env.s("complex_math_string", get_complex_math_string());
@@ -773,15 +778,7 @@ std::string generate_code(
     env.s("complex_math_string", "");
   }
   if (f_inputs_type == "std::complex<at::Half>" ||
-      result_type == "std::complex<at::Half>") {
-    // complex<Half> depends on complex<T> and Half dtypes.
-    env.s("traits_string", get_traits_string());
-    env.s("half_string", jiterator_half_support_literal);
-    env.s("complex_body_string", get_complex_body_string());
-    env.s("complex_math_string", get_complex_math_string());
-
-    env.s("complex_half_body_string", get_complex_half_body_string());
-  } else if (dynamic_casting) {
+      result_type == "std::complex<at::Half>" || dynamic_casting) {
     env.s("complex_half_body_string", get_complex_half_body_string());
   } else {
     env.s("complex_half_body_string", "");
```

Thanks!

Reply:
Yup! I believe the logic is easier to follow this way, as every field is set in just one place. Also, adding a comment for why is `get_complex_half_body_string` needed in dynamic casting similar to the one that you have before complex would also be great :)

Reply:
Thanks! On it :)